File: crates/viben-core/src/gateway/routes/terminal.rs
//! Terminal WebSocket Route
//!
//! Provides WebSocket endpoint for terminal emulation using PTY sessions.
//! Supports input, resize, and output streaming with base64 encoding.

use std::path::PathBuf;

use axum::{
    Router,
    extract::{Query, State, ws::{Message, WebSocket, WebSocketUpgrade}},
    response::IntoResponse,
    routing::get,
};
use base64::{Engine, engine::general_purpose::STANDARD as BASE64};
use futures_util::{SinkExt, StreamExt};
use serde::{Deserialize, Serialize};

use crate::gateway::AppState;
use crate::gateway::error::GatewayError;

/// Query parameters for terminal WebSocket connection
#[derive(Debug, Deserialize)]
pub struct TerminalQuery {
    /// Working directory for the terminal session
    pub cwd: Option<String>,
    /// Number of columns (default: 80)
    #[serde(default = "default_cols")]
    pub cols: u16,
    /// Number of rows (default: 24)
    #[serde(default = "default_rows")]
    pub rows: u16,
}

fn default_cols() -> u16 {
    80
}

fn default_rows() -> u16 {
    24
}

/// Commands sent from client to terminal
#[derive(Debug, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum TerminalCommand {
    /// Input data (base64 encoded)
    Input { data: String },
    /// Resize terminal
    Resize { cols: u16, rows: u16 },
}

/// Messages sent from terminal to client
#[derive(Debug, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum TerminalMessage {
    /// Output data (base64 encoded)
    Output { data: String },
    /// Error message
    Error { message: String },
    /// Session connected
    Connected { session_id: String },
}

/// Terminal WebSocket upgrade handler
pub async fn terminal_ws(
    ws: WebSocketUpgrade,
    State(state): State<AppState>,
    Query(query): Query<TerminalQuery>,
) -> Result<impl IntoResponse, GatewayError> {
    // Determine working directory
    let working_dir = if let Some(cwd) = &query.cwd {
        let path = PathBuf::from(cwd);
        if !path.exists() {
            return Err(GatewayError::BadRequest(format!(
                "Working directory does not exist: {}",
                cwd
            )));
        }
        path
    } else {
        std::env::current_dir().map_err(|e| {
            GatewayError::Internal(format!("Failed to get current directory: {}", e))
        })?
    };

    Ok(ws.on_upgrade(move |socket| {
        handle_terminal_ws(socket, state, working_dir, query.cols, query.rows)
    }))
}

/// Handle terminal WebSocket connection
async fn handle_terminal_ws(
    socket: WebSocket,
    state: AppState,
    working_dir: PathBuf,
    cols: u16,
    rows: u16,
) {
    // Create PTY session
    let (session_id, mut output_rx) = match state
        .pty
        .create_session(working_dir.clone(), cols, rows)
        .await
    {
        Ok(result) => result,
        Err(e) => {
            tracing::error!("Failed to create PTY session: {}", e);
            let _ = send_error(socket, &e.to_string()).await;
            return;
        }
    };

    let (mut ws_sender, mut ws_receiver) = socket.split();

    // Send connected message
    let connected_msg = TerminalMessage::Connected {
        session_id: session_id.to_string(),
    };
    if let Ok(json) = serde_json::to_string(&connected_msg) {
        if ws_sender.send(Message::Text(json)).await.is_err() {
            let _ = state.pty.close_session(session_id).await;
            return;
        }
    }

    let pty_service = state.pty.clone();
    let session_id_for_input = session_id;

    // Task to forward PTY output to WebSocket
    let output_task = tokio::spawn(async move {
        while let Some(data) = output_rx.recv().await {
            let msg = TerminalMessage::Output {
                data: BASE64.encode(&data),
            };
            let json = match serde_json::to_string(&msg) {
                Ok(j) => j,
                Err(_) => continue,
            };
            if ws_sender.send(Message::Text(json)).await.is_err() {
                break;
            }
        }
        ws_sender
    });

    // Handle incoming WebSocket messages
    while let Some(Ok(msg)) = ws_receiver.next().await {
        match msg {
            Message::Text(text) => {
                if let Ok(cmd) = serde_json::from_str::<TerminalCommand>(&text) {
                    match cmd {
                        TerminalCommand::Input { data } => {
                            if let Ok(bytes) = BASE64.decode(&data) {
                                if let Err(e) = pty_service.write(session_id_for_input, &bytes).await {
                                    tracing::warn!("Failed to write to PTY: {}", e);
                                }
                            }
                        }
                        TerminalCommand::Resize { cols, rows } => {
                            if let Err(e) = pty_service.resize(session_id_for_input, cols, rows).await {
                                tracing::warn!("Failed to resize PTY: {}", e);
                            }
                        }
                    }
                }
            }
            Message::Binary(data) => {
                // Support raw binary input as well
                if let Err(e) = pty_service.write(session_id_for_input, &data).await {
                    tracing::warn!("Failed to write binary to PTY: {}", e);
                }
            }
            Message::Close(_) => break,
            _ => {}
        }
    }

    // Cleanup
    let _ = state.pty.close_session(session_id).await;
    output_task.abort();

    tracing::info!("Terminal WebSocket connection closed for session {}", session_id);
}

/// Send error message and close WebSocket
async fn send_error(mut socket: WebSocket, message: &str) -> Result<(), axum::Error> {
    let msg = TerminalMessage::Error {
        message: message.to_string(),
    };
    let json = serde_json::to_string(&msg).unwrap_or_default();
    socket.send(Message::Text(json)).await?;
    socket.close().await?;
    Ok(())
}

/// Create terminal router
pub fn router() -> Router<AppState> {
    Router::new().route("/terminal/ws", get(terminal_ws))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_terminal_command_deserialize() {
        // Test input command
        let input_json = r#"{"type": "input", "data": "aGVsbG8="}"#;
        let cmd: TerminalCommand = serde_json::from_str(input_json).unwrap();
        match cmd {
            TerminalCommand::Input { data } => {
                assert_eq!(data, "aGVsbG8=");
                let decoded = BASE64.decode(&data).unwrap();
                assert_eq!(decoded, b"hello");
            }
            _ => panic!("Expected Input command"),
        }

        // Test resize command
        let resize_json = r#"{"type": "resize", "cols": 120, "rows": 40}"#;
        let cmd: TerminalCommand = serde_json::from_str(resize_json).unwrap();
        match cmd {
            TerminalCommand::Resize { cols, rows } => {
                assert_eq!(cols, 120);
                assert_eq!(rows, 40);
            }
            _ => panic!("Expected Resize command"),
        }
    }

    #[test]
    fn test_terminal_message_serialize() {
        // Test output message
        let msg = TerminalMessage::Output {
            data: BASE64.encode(b"hello"),
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains("output"));
        assert!(json.contains("aGVsbG8="));

        // Test connected message
        let msg = TerminalMessage::Connected {
            session_id: "test-123".to_string(),
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains("connected"));
        assert!(json.contains("test-123"));

        // Test error message
        let msg = TerminalMessage::Error {
            message: "Test error".to_string(),
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains("error"));
        assert!(json.contains("Test error"));
    }

    #[test]
    fn test_terminal_query_defaults() {
        let query: TerminalQuery = serde_json::from_str("{}").unwrap();
        assert_eq!(query.cols, 80);
        assert_eq!(query.rows, 24);
        assert!(query.cwd.is_none());
    }
}

File: packages/core/src/gateway/routes/terminal.ts
/**
 * Terminal WebSocket Route
 *
 * Provides WebSocket endpoint for PTY terminal emulation.
 * Supports input, resize, and output streaming with base64 encoding.
 *
 * Note: This module requires node-pty as an optional dependency.
 * If node-pty is not available, terminal routes will be disabled.
 */
import type { FastifyInstance } from "fastify";
import type { AppState } from "../state";
import { randomUUID } from "crypto";
import * as fs from "fs";
import * as os from "os";

/**
 * Query parameters for terminal WebSocket connection
 */
interface TerminalQuery {
  /** Working directory for the terminal session */
  cwd?: string;
  /** Number of columns (default: 80) */
  cols?: number;
  /** Number of rows (default: 24) */
  rows?: number;
  /** Shell to use (default: system default) */
  shell?: string;
}

/**
 * Commands sent from client to terminal
 */
interface TerminalCommand {
  type: "input" | "resize";
  /** Base64 encoded input data (for type: "input") */
  data?: string;
  /** Number of columns (for type: "resize") */
  cols?: number;
  /** Number of rows (for type: "resize") */
  rows?: number;
}

/**
 * Messages sent from terminal to client
 */
interface TerminalMessage {
  type: "output" | "exit" | "error" | "connected";
  /** Base64 encoded output data (for type: "output") */
  data?: string;
  /** Exit code (for type: "exit") */
  code?: number;
  /** Error message (for type: "error") */
  message?: string;
  /** Session ID (for type: "connected") */
  sessionId?: string;
}

/**
 * PTY session information
 */
interface PtySession {
  id: string;
  pty: IPty;
  createdAt: Date;
}

/**
 * node-pty IPty interface (minimal subset we use)
 */
interface IPty {
  pid: number;
  cols: number;
  rows: number;
  process: string;
  onData: (callback: (data: string) => void) => { dispose: () => void };
  onExit: (callback: (e: { exitCode: number; signal?: number }) => void) => { dispose: () => void };
  write: (data: string) => void;
  resize: (cols: number, rows: number) => void;
  kill: (signal?: string) => void;
}

/**
 * node-pty spawn function interface
 */
interface NodePtyModule {
  spawn: (
    shell: string,
    args: string[],
    options: {
      name?: string;
      cols?: number;
      rows?: number;
      cwd?: string;
      env?: Record<string, string>;
    }
  ) => IPty;
}

/** Active PTY sessions */
const sessions = new Map<string, PtySession>();

/**
 * Get the default shell for the current platform
 */
function getDefaultShell(): string {
  if (os.platform() === "win32") {
    return process.env.COMSPEC || "cmd.exe";
  }
  return process.env.SHELL || "/bin/bash";
}

/**
 * Base64 encode a string
 */
function base64Encode(str: string): string {
  return Buffer.from(str, "utf-8").toString("base64");
}

/**
 * Base64 decode a string
 */
function base64Decode(str: string): string {
  return Buffer.from(str, "base64").toString("utf-8");
}

/**
 * Register terminal WebSocket routes
 *
 * Note: @fastify/websocket plugin must be registered at the gateway level before calling this function.
 * This prevents ERR_HTTP_SOCKET_ASSIGNED errors from multiple registrations.
 *
 * @param fastify - Fastify instance
 * @param _state - Application state (unused but kept for consistency)
 */
export function registerTerminalRoutes(fastify: FastifyInstance, _state?: AppState): void {
  // Check if websocket plugin is registered
  if (!fastify.hasDecorator("websocketServer")) {
    console.warn("[Terminal] @fastify/websocket not registered, terminal WebSocket routes disabled");
    return;
  }

  // Register WebSocket route dynamically if node-pty is available
  fastify.register(async (instance) => {
    let nodePty: NodePtyModule | null = null;

    // Try to load node-pty
    try {
      nodePty = (await import("node-pty")) as unknown as NodePtyModule;
    } catch {
      console.warn("[Terminal] node-pty not available, terminal WebSocket routes disabled");
      console.warn("[Terminal] Install node-pty to enable: npm install node-pty");
      return;
    }

    // WebSocket endpoint for terminal
    instance.get<{
      Querystring: TerminalQuery;
    }>("/ws/terminal", { websocket: true }, (socket, req) => {
      const query = req.query;
      const cols = query.cols ?? 80;
      const rows = query.rows ?? 24;
      const shell = query.shell ?? getDefaultShell();
      const cwd = query.cwd ?? process.cwd();

      // Validate working directory
      if (query.cwd && !fs.existsSync(query.cwd)) {
        const errorMsg: TerminalMessage = {
          type: "error",
          message: `Working directory does not exist: ${query.cwd}`,
        };
        socket.send(JSON.stringify(errorMsg));
        socket.close();
        return;
      }

      // Generate session ID
      const sessionId = randomUUID();

      // Create PTY process
      let pty: IPty;
      try {
        pty = nodePty!.spawn(shell, [], {
          name: "xterm-256color",
          cols,
          rows,
          cwd,
          env: process.env as Record<string, string>,
        });
      } catch (err) {
        const errorMsg: TerminalMessage = {
          type: "error",
          message: `Failed to spawn PTY: ${err instanceof Error ? err.message : String(err)}`,
        };
        socket.send(JSON.stringify(errorMsg));
        socket.close();
        return;
      }

      // Store session
      const session: PtySession = {
        id: sessionId,
        pty,
        createdAt: new Date(),
      };
      sessions.set(sessionId, session);

      // Send connected message
      const connectedMsg: TerminalMessage = {
        type: "connected",
        sessionId,
      };
      socket.send(JSON.stringify(connectedMsg));

      // Handle PTY output
      const dataDisposable = pty.onData((data: string) => {
        const outputMsg: TerminalMessage = {
          type: "output",
          data: base64Encode(data),
        };
        try {
          socket.send(JSON.stringify(outputMsg));
        } catch {
          // Socket might be closed
        }
      });

      // Handle PTY exit
      const exitDisposable = pty.onExit((e: { exitCode: number; signal?: number }) => {
        const exitMsg: TerminalMessage = {
          type: "exit",
          code: e.exitCode,
        };
        try {
          socket.send(JSON.stringify(exitMsg));
          socket.close();
        } catch {
          // Socket might already be closed
        }
        cleanup();
      });

      // Cleanup function
      const cleanup = () => {
        dataDisposable.dispose();
        exitDisposable.dispose();
        sessions.delete(sessionId);
        try {
          pty.kill();
        } catch {
          // PTY might already be dead
        }
      };

      // Handle incoming messages
      socket.on("message", (data: Buffer) => {
        try {
          const msg = JSON.parse(data.toString()) as TerminalCommand;

          switch (msg.type) {
            case "input": {
              if (msg.data) {
                const decoded = base64Decode(msg.data);
                pty.write(decoded);
              }
              break;
            }

            case "resize": {
              const newCols = msg.cols ?? cols;
              const newRows = msg.rows ?? rows;
              if (newCols > 0 && newRows > 0) {
                pty.resize(newCols, newRows);
              }
              break;
            }
          }
        } catch {
          const errorMsg: TerminalMessage = {
            type: "error",
            message: "Failed to parse message",
          };
          socket.send(JSON.stringify(errorMsg));
        }
      });

      // Handle WebSocket close
      socket.on("close", () => {
        cleanup();
        console.log(`[Terminal] WebSocket connection closed for session ${sessionId}`);
      });

      // Handle WebSocket error
      socket.on("error", (err) => {
        console.error(`[Terminal] WebSocket error for session ${sessionId}:`, err);
        cleanup();
      });
    });

    console.log("[Terminal] Terminal WebSocket routes registered at /ws/terminal");
  });
}

/**
 * Get active terminal sessions count
 */
export function getActiveSessionCount(): number {
  return sessions.size;
}

/**
 * Kill all active terminal sessions
 */
export function killAllSessions(): void {
  for (const [sessionId, session] of sessions) {
    try {
      session.pty.kill();
    } catch {
      // Ignore errors
    }
    sessions.delete(sessionId);
  }
}
